File: algorithms/C/Count Unhappy Friends/Count Unhappy Friends.py
class Solution:
    def unhappyFriends(self, n: int, preferences: List[List[int]], pairs: List[List[int]]) -> int:
        dd = {}
        
        for i,x in pairs:
            dd[i] = preferences[i][:preferences[i].index(x)]
            dd[x] = preferences[x][:preferences[x].index(i)]
        
        ans = 0
            
        for i in dd:
            for x in dd[i]:
                if i in dd[x]:
                    ans += 1
                    break
        
        return ans

File: scripts/algorithms/C/Count Unhappy Friends/Count Unhappy Friends.java
class Solution {
  public int unhappyFriends(int n, int[][] preferences, int[][] pairs) {
    int[][] rankings = new int[n][n]; // smaller the value, higher the preference
    int[] pairedWith = new int[n];
    for (int i = 0; i < n; i++) {
      for (int rank = 0; rank < n - 1; rank++) {
        int j = preferences[i][rank];
        rankings[i][j] = rank; // person "i" views person "j" with rank
      }
    }
    int unhappy = 0;
    for (int[] pair : pairs) {
      int a = pair[0], b = pair[1];
      pairedWith[a] = b;
      pairedWith[b] = a;
    }
    for (int a = 0; a < n; a++) {
      // "a" prefers someone else
      if (rankings[a][pairedWith[a]] != 0) {
        for (int b = 0; b < n; b++) {
          // "b" prefers to be with "a" over their current partner
          // "a" prefers to be with "b" over their current partner
          if (b != a
              && rankings[b][a] < rankings[b][pairedWith[b]]
              && rankings[a][b] < rankings[a][pairedWith[a]]) {
            unhappy++;
            break;
          }
        }
      }
    }
    return unhappy;
  }
}
